### DISCUSSION TITLE
'LibtorchTorchScriptNeuralNetTest' is not a registered object.

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#26828] [Q&A General]
Hi,
This occurs when i intend to execute a simple test from the repository that loads a torch script file. The libtorch installations were done separately from moose and are linked correctly. 

The error log has the following form.
```
*** ERROR ***
A 'LibtorchTorchScriptNeuralNetTest' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.^[[39m

Abort(1) on node 6 (rank 6 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 6
```
The test case has the following input commands.
```
[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 1
  []
[]

[VectorPostprocessors]
  [test]
    type = LibtorchTorchScriptNeuralNetTest
    filename = "my_net.pt"
  []
[]

[Problem]
  solve = false
[]

[Executioner]
  type = Steady
[]

[Outputs]
  csv = true
[]

```
STOCHASTIC_TOOLS module is included for building and should compile and link correctly. 
Look forward to your suggestions.
Kind regards,
Arun

**Comment by GiudGiud** [comment]
Hello

Including Stochastic tools is not enough. 

You have to go on the stochastic tools module webpage to look for the Libtorch installation instructions

Guillaume

**Reply by grmnptr** [reply]
https://mooseframework.inl.gov/getting_started/installation/install_libtorch.html

**Reply by abarun22** [reply]
I am pretty sure libtorch has been installed fine and configured while building moose.
It should run fine for this simple case i believe.
Could be a tricky thing, but definitely solvable.

**Reply by GiudGiud** [reply]
what does `ldd stochastic_tools-opt` (on linux) return?

**Comment by abarun22** [comment]
I see libtorch linked against this executable making it look less
suspicious. I am kind of clueless as to what really is required for the fix.

```

***@***.*** stochastic_tools]$ ldd stochastic_tools-opt
        linux-vdso.so.1 (0x00007ffc86bb8000)
        libstochastic_tools_test-opt.so.0 =>
/home/ir-bala2/moose_dev/moose/modules/stochastic_tools/test/lib/libstochastic_tools_test-opt.so.0
(0x00001547eb778000)
        libmoose-opt.so.0 =>
/home/ir-bala2/moose_dev/moose/framework/libmoose-opt.so.0
(0x00001547e7a7e000)
        libpng16.so.16 => /lib64/libpng16.so.16 (0x00001547e7849000)
        libpcre-opt.so.0 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/pcre/libpcre-opt.so.0
(0x00001547e7620000)
        libhit-opt.so.0 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/hit/libhit-opt.so.0
(0x00001547e7389000)
        libstochastic_tools-opt.so.0 =>
/home/ir-bala2/moose_dev/moose/modules/stochastic_tools/lib/libstochastic_tools-opt.so.0
(0x00001547e677a000)
        libwasphit.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwasphit.so.04
(0x00001547e652f000)
        libwasplsp.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwasplsp.so.04
(0x00001547e62d0000)
        libwaspson.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwaspson.so.04
(0x00001547e607f000)
        libwaspcore.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwaspcore.so.04
(0x00001547e5e58000)
        libwaspsiren.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwaspsiren.so.04
(0x00001547e5c3a000)
        libwasphalite.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwasphalite.so.04
(0x00001547e5999000)
        libwaspjson.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwaspjson.so.04
(0x00001547e5751000)
        libwasphive.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwasphive.so.04
(0x00001547e5547000)
        libwaspddi.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwaspddi.so.04
(0x00001547e5327000)
        libwaspexpr.so.04 =>
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib/libwaspexpr.so.04
(0x00001547e50ee000)
        libtorch.so =>
/home/ir-bala2/rds/rds-ukaea-ap001/arunpb/Tools/pytorch/pytorch-install/lib/libtorch.so
(0x00001547e4eec000)
        libmesh_opt.so.0 =>
/home/ir-bala2/moose_dev/moose/scripts/../libmesh/installed/lib/libmesh_opt.so.0
(0x00001547e2bca000)
        libnetcdf.so.13 =>
/home/ir-bala2/moose_dev/moose/scripts/../libmesh/installed/lib/libnetcdf.so.13
(0x00001547e28cc000)
        libtimpi_opt.so.11 =>
/home/ir-bala2/moose_dev/moose/scripts/../libmesh/installed/lib/libtimpi_opt.so.11
(0x00001547e26bb000)
        libslepc.so.3.20 =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libslepc.so.3.20
(0x00001547e2008000)
        libpetsc.so.3.20 =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libpetsc.so.3.20
(0x00001547dfa5b000)
        libHYPRE-2.29.0.so =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libHYPRE-2.29.0.so
(0x00001547df0e4000)
        libstrumpack.so.7.1 =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libstrumpack.so.7.1
(0x00001547dd845000)
        libscalapack.so.2.2 =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libscalapack.so.2.2
(0x00001547dcd47000)
        libsuperlu_dist.so.8 =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libsuperlu_dist.so.8
(0x00001547dc821000)
        libparmetis.so =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libparmetis.so
(0x00001547dc5c5000)
        libmetis.so =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libmetis.so
(0x00001547dc335000)
        libhdf5_hl.so.200 =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libhdf5_hl.so.200
(0x00001547dc10e000)
        libhdf5.so.200 =>
/home/ir-bala2/moose_dev/moose/petsc/arch-moose/lib/libhdf5.so.200
(0x00001547db9f6000)
        libz.so.1 => /lib64/libz.so.1 (0x00001547db7de000)
        libX11.so.6 => /lib64/libX11.so.6 (0x00001547db49a000)
        libifport.so.5 =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libifport.so.5
(0x00001547db268000)
        libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00001547db027000)
        libmpicxx.so.12 =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-mpi-2021.6.0/intel-2021.6.0/guxuvcpmykplbrr2e3af2yd7njqhau5e/mpi/2021.6.0//lib/libmpicxx.so.12
(0x00001547dae07000)
        libmpifort.so.12 =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-mpi-2021.6.0/intel-2021.6.0/guxuvcpmykplbrr2e3af2yd7njqhau5e/mpi/2021.6.0//lib/libmpifort.so.12
(0x00001547daa53000)
        libmpi.so.12 =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-mpi-2021.6.0/intel-2021.6.0/guxuvcpmykplbrr2e3af2yd7njqhau5e/mpi/2021.6.0//lib/release/libmpi.so.12
(0x00001547d920b000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00001547d9007000)
        librt.so.1 => /lib64/librt.so.1 (0x00001547d8dff000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00001547d8bdf000)
        libimf.so =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libimf.so
(0x00001547d854a000)
        libsvml.so =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libsvml.so
(0x00001547d650c000)
        libirng.so =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libirng.so
(0x00001547d61a0000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00001547d5e0b000)
        libm.so.6 => /lib64/libm.so.6 (0x00001547d5a89000)
        libcilkrts.so.5 =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libcilkrts.so.5
(0x00001547d5847000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00001547d562f000)
        libirc.so =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libirc.so
(0x00001547d53b4000)
        libtirpc.so.3 => /lib64/libtirpc.so.3 (0x00001547d5180000)
        libiomp5.so =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libiomp5.so
(0x00001547d4d3c000)
        libc.so.6 => /lib64/libc.so.6 (0x00001547d4977000)
        libintlc.so.5 =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libintlc.so.5
(0x00001547d46fc000)
        /lib64/ld-linux-x86-64.so.2 (0x00001547eba92000)
        libtorch_cpu.so =>
/home/ir-bala2/rds/rds-ukaea-ap001/arunpb/Tools/pytorch/pytorch-install/lib/libtorch_cpu.so
(0x00001547c967a000)
        libtorch_cuda.so =>
/home/ir-bala2/rds/rds-ukaea-ap001/arunpb/Tools/pytorch/pytorch-install/lib/libtorch_cuda.so
(0x000015479723d000)
        libifcoremt.so.5 =>
/usr/local/software/spack/spack-views/rocky8-icelake-20220710/intel-oneapi-compilers-2022.1.0/gcc-11.3.0/b6zld2mz7cid27yloxznoidymd7vywwz/compiler/2022.1.0/linux/compiler/lib/intel64_lin/libifcoremt.so.5
(0x0000154797094000)
        libxcb.so.1 => /lib64/libxcb.so.1 (0x0000154796e6b000)
        libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2
(0x0000154796c16000)
        libkrb5.so.3 => /lib64/libkrb5.so.3 (0x000015479692b000)
        libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x0000154796714000)
        libcom_err.so.2 => /lib64/libcom_err.so.2 (0x0000154796510000)
        libc10.so =>
/home/ir-bala2/rds/rds-ukaea-ap001/arunpb/Tools/pytorch/pytorch-install/lib/libc10.so
(0x0000154796239000)
        libopenblas.so.0 => /lib64/libopenblas.so.0 (0x0000154793f8a000)
        libomp.so => /lib64/libomp.so (0x0000154793c8e000)
        libcupti.so.11.4 =>
/usr/local/software/cuda/11.4/extras/CUPTI/lib64/libcupti.so.11.4
(0x0000154793360000)
        libcudart.so.11.0 =>
/usr/local/software/cuda/11.4/lib64/libcudart.so.11.0 (0x00001547930be000)
        libc10_cuda.so =>
/home/ir-bala2/rds/rds-ukaea-ap001/arunpb/Tools/pytorch/pytorch-install/lib/libc10_cuda.so
(0x0000154792e35000)
        libcusparse.so.11 =>
/usr/local/software/cuda/11.4/lib64/libcusparse.so.11 (0x0000154784a4a000)
        libcufft.so.10 =>
/usr/local/software/cuda/11.4/lib64/libcufft.so.10 (0x000015476edca000)
        libnvToolsExt.so.1 =>
/usr/local/software/cuda/11.4/lib64/libnvToolsExt.so.1 (0x000015476ebc0000)
        libcurand.so.10 =>
/usr/local/software/cuda/11.4/lib64/libcurand.so.10 (0x0000154769411000)
        libcublas.so.11 =>
/usr/local/software/cuda/11.4/lib64/libcublas.so.11 (0x00001547604d8000)
        libcublasLt.so.11 =>
/usr/local/software/cuda/11.4/lib64/libcublasLt.so.11 (0x000015474e3c9000)
        libXau.so.6 => /lib64/libXau.so.6 (0x000015474e1c5000)
        libkrb5support.so.0 => /lib64/libkrb5support.so.0
(0x000015474dfb4000)
        libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x000015474ddb0000)
        libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x000015474d8c6000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x000015474d6ae000)
        libnuma.so.1 => /lib64/libnuma.so.1 (0x000015474d4a2000)
        libgfortran.so.5 => /lib64/libgfortran.so.5 (0x000015474d023000)
        libutil.so.1 => /lib64/libutil.so.1 (0x000015474ce1f000)
        libselinux.so.1 => /lib64/libselinux.so.1 (0x000015474cbf4000)
        libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x000015474c970000)

```

**Reply by grmnptr** [reply]
When you run this in the stochastic tools folder: `run_tests -j10 --re=libtorch` does it skip or successfully finish the tests?

**Comment by GiudGiud** [comment]
what is this LibtorchTorchScriptNeuralNetTest object?

where did you see it?
It looks like a test object to me

**Reply by grmnptr** [reply]
Aaa yes, it is a framework test object. It won't compile in the stochastic tools app.

**Reply by grmnptr** [reply]
To run the test, you will need to go to `moose/test` and compile that and then use `moose_test-opt` to run your file.

**Reply by abarun22** [reply]
Here are the test results.

[ir-bala2@login-q-2 stochastic_tools]$ ./run_tests -j10 --re=libtorch
examples/surrogates/cross_validation.cv/cv_no_libtorch ................... [LIBTORCH!=FALSE, no deepdiff] SKIP
test:transfers/libtorch_nn_transfer.torch-parameters ...................................... [no deepdiff] SKIP
test:functions/drl_reward.libtorch_drl_reward_linear: Working Directory: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/test/tests/functions/drl_reward
test:functions/drl_reward.libtorch_drl_reward_linear: Running command: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt -i drl_reward.i Outputs/file_base='drl_reward_linear' --error --error-override --no-gdb-backtrace
test:functions/drl_reward.libtorch_drl_reward_linear: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt: symbol lookup error: /home/ir-bala2/moose_dev/moose/framework/libmoose-opt.so.0: undefined symbol: _ZNK5torch2nn6Module5cloneERKN3c108optionalINS2_6DeviceEEE
test:functions/drl_reward.libtorch_drl_reward_linear:
test:functions/drl_reward.libtorch_drl_reward_linear: ################################################################################
test:functions/drl_reward.libtorch_drl_reward_linear: Tester failed, reason: CRASH
test:functions/drl_reward.libtorch_drl_reward_linear:
test:functions/drl_reward.libtorch_drl_reward_linear .......................................... FAILED (CRASH)
examples/libtorch_drl_control.train: Working Directory: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/examples/libtorch_drl_control
examples/libtorch_drl_control.train: Running command: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt -i libtorch_drl_control_trainer.i Executioner/num_steps=1 Trainers/nn_trainer/update_frequency=1 Trainers/nn_trainer/num_epochs=100 Outputs/interval=1 Outputs/file_base='train_out' --error --error-override --no-gdb-backtrace
examples/libtorch_drl_control.train: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt: symbol lookup error: /home/ir-bala2/moose_dev/moose/framework/libmoose-opt.so.0: undefined symbol: _ZNK5torch2nn6Module5cloneERKN3c108optionalINS2_6DeviceEEE
examples/libtorch_drl_control.train:
examples/libtorch_drl_control.train: ################################################################################
examples/libtorch_drl_control.train: Tester failed, reason: CRASH
examples/libtorch_drl_control.train:
examples/libtorch_drl_control.train ........................................................... FAILED (CRASH)
test:controls/libtorch_drl_control.read-parameters: Working Directory: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/test/tests/controls/libtorch_drl_control
test:controls/libtorch_drl_control.read-parameters: Running command: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt -i libtorch_drl_control.i Outputs/file_base='parameter_read' --error --error-override --no-gdb-backtrace
test:controls/libtorch_drl_control.read-parameters: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt: symbol lookup error: /home/ir-bala2/moose_dev/moose/framework/libmoose-opt.so.0: undefined symbol: _ZNK5torch2nn6Module5cloneERKN3c108optionalINS2_6DeviceEEE
test:controls/libtorch_drl_control.read-parameters:
test:controls/libtorch_drl_control.read-parameters: ################################################################################
test:controls/libtorch_drl_control.read-parameters: Tester failed, reason: CRASH
test:controls/libtorch_drl_control.read-parameters:
test:controls/libtorch_drl_control.read-parameters ............................................ FAILED (CRASH)
test:surrogates/libtorch_nn.train_and_evaluate: Working Directory: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/test/tests/surrogates/libtorch_nn
test:surrogates/libtorch_nn.train_and_evaluate: Running command: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt -i train_and_evaluate.i Trainers/train/read_from_file=false --allow-test-objects --error --error-override --no-gdb-backtrace
test:surrogates/libtorch_nn.train_and_evaluate: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt: symbol lookup error: /home/ir-bala2/moose_dev/moose/framework/libmoose-opt.so.0: undefined symbol: _ZNK5torch2nn6Module5cloneERKN3c108optionalINS2_6DeviceEEE
test:surrogates/libtorch_nn.train_and_evaluate:
test:surrogates/libtorch_nn.train_and_evaluate: ################################################################################
test:surrogates/libtorch_nn.train_and_evaluate: Tester failed, reason: CRASH
test:surrogates/libtorch_nn.train_and_evaluate:
test:surrogates/libtorch_nn.train_and_evaluate ................................................ FAILED (CRASH)
test:surrogates/libtorch_nn.retrain ................................................ [skipped dependency] SKIP
test:surrogates/libtorch_nn.train: Working Directory: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/test/tests/surrogates/libtorch_nn
test:surrogates/libtorch_nn.train: Running command: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt -i train.i Trainers/train/read_from_file=false --allow-test-objects --error --error-override --no-gdb-backtrace
test:surrogates/libtorch_nn.train: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt: symbol lookup error: /home/ir-bala2/moose_dev/moose/framework/libmoose-opt.so.0: undefined symbol: _ZNK5torch2nn6Module5cloneERKN3c108optionalINS2_6DeviceEEE
test:surrogates/libtorch_nn.train:
test:surrogates/libtorch_nn.train: ################################################################################
test:surrogates/libtorch_nn.train: Tester failed, reason: CRASH
test:surrogates/libtorch_nn.train:
test:surrogates/libtorch_nn.train ............................................................. FAILED (CRASH)
test:surrogates/libtorch_nn.train-with-rel-tol ..................................... [skipped dependency] SKIP
test:surrogates/libtorch_nn.evaluate ............................................... [skipped dependency] SKIP
test:controls/libtorch_drl_control.without-nn: Working Directory: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/test/tests/controls/libtorch_drl_control
test:controls/libtorch_drl_control.without-nn: Running command: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt -i libtorch_drl_control.i Outputs/file_base='run_wo_nn' Controls/inactive='src_control' Postprocessors/left_flux/control_name='src_control_empty' Postprocessors/log_prob_left_flux/control_name='src_control_empty' --error --error-override --no-gdb-backtrace
test:controls/libtorch_drl_control.without-nn: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt: symbol lookup error: /home/ir-bala2/moose_dev/moose/framework/libmoose-opt.so.0: undefined symbol: _ZNK5torch2nn6Module5cloneERKN3c108optionalINS2_6DeviceEEE
test:controls/libtorch_drl_control.without-nn:
test:controls/libtorch_drl_control.without-nn: ################################################################################
test:controls/libtorch_drl_control.without-nn: Tester failed, reason: CRASH
test:controls/libtorch_drl_control.without-nn:
test:controls/libtorch_drl_control.without-nn ................................................. FAILED (CRASH)
test:surrogates/libtorch_nn.train-standardized: Working Directory: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/test/tests/surrogates/libtorch_nn
test:surrogates/libtorch_nn.train-standardized: Running command: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt -i train.i Trainers/train/read_from_file=false Trainers/train/standardize_input=true Trainers/train/standardize_output=true Outputs/file_base=standardized --allow-test-objects --error --error-override --no-gdb-backtrace
test:surrogates/libtorch_nn.train-standardized: /home/ir-bala2/moose_dev/moose/modules/stochastic_tools/stochastic_tools-opt: symbol lookup error: /home/ir-bala2/moose_dev/moose/framework/libmoose-opt.so.0: undefined symbol: _ZNK5torch2nn6Module5cloneERKN3c108optionalINS2_6DeviceEEE
test:surrogates/libtorch_nn.train-standardized:
test:surrogates/libtorch_nn.train-standardized: ################################################################################
test:surrogates/libtorch_nn.train-standardized: Tester failed, reason: CRASH
test:surrogates/libtorch_nn.train-standardized:
test:surrogates/libtorch_nn.train-standardized ................................................ FAILED (CRASH)
test:surrogates/libtorch_nn.evaluate-standardized .................................. [skipped dependency] SKIP


Final Test Results:
--------------------------------------------------------------------------------------------------------------
examples/surrogates/cross_validation.cv/cv_no_libtorch ................... [LIBTORCH!=FALSE, no deepdiff] SKIP
test:transfers/libtorch_nn_transfer.torch-parameters ...................................... [no deepdiff] SKIP
test:surrogates/libtorch_nn.retrain ................................................ [skipped dependency] SKIP
test:surrogates/libtorch_nn.train-with-rel-tol ..................................... [skipped dependency] SKIP
test:surrogates/libtorch_nn.evaluate ............................................... [skipped dependency] SKIP
test:surrogates/libtorch_nn.evaluate-standardized .................................. [skipped dependency] SKIP
test:functions/drl_reward.libtorch_drl_reward_linear .......................................... FAILED (CRASH)
examples/libtorch_drl_control.train ........................................................... FAILED (CRASH)
test:controls/libtorch_drl_control.read-parameters ............................................ FAILED (CRASH)
test:surrogates/libtorch_nn.train_and_evaluate ................................................ FAILED (CRASH)
test:surrogates/libtorch_nn.train ............................................................. FAILED (CRASH)
test:controls/libtorch_drl_control.without-nn ................................................. FAILED (CRASH)
test:surrogates/libtorch_nn.train-standardized ................................................ FAILED (CRASH)
--------------------------------------------------------------------------------------------------------------
Ran 7 tests in 30.2 seconds. Average test time 8.1 seconds, maximum test time 10.1 seconds.
0 passed, 6 skipped, 0 pending, 7 FAILED

Clearly experiencing some issues.

**Reply by abarun22** [reply]
This has gone a complete reversal, when i re-compiled the test object.
(py3venv) [ir-bala2@login-p-3 test]$ make -j10 install

***ERROR***
WASP does not seem to be available.
Make sure to either run scripts/update_and_rebuild_wasp.sh in your MOOSE directory,
or set WASP_DIR to a valid WASP install
make: *** [wasp_submodule_status] Error 1

I am sure the LD_LIBRARY_PATH and WASP_DIR are pointing to the correct paths contrib/wasp/install/lib and contrib/wasp/install respectively. I have not changed anything in the repository, so no clue as what's going wrong.
Any advice here will be highly appreciated.
Kind regards,
Arun

**Reply by GiudGiud** [reply]
you forgot to conda activate moose

**Reply by abarun22** [reply]
Do you recommend this to be done within a conda environment? 
The command returns the following.

```
(base) [ir-bala2@login-p-1 moose]$ conda activate moose
Could not find conda environment: moose
You can list all discoverable environments with `conda info --envs`.

```

**Reply by GiudGiud** [reply]
So you are on a cluster then?
go in moose/scripts and install wasp then, as the error message said

**Comment by abarun22** [comment]
Yes, I am working from within a cluster.
Wasp has already been installed as you suggested from the update script in
the repo.

(base) ***@***.*** test]$ ll
/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install/lib
total 2792
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     20 Feb 15 12:02 libwaspcore.so.04 ->
libwaspcore.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 194576 Feb 15 12:02 libwaspcore.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     19 Feb 15 12:02 libwaspddi.so.04 ->
libwaspddi.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 161088 Feb 15 12:02 libwaspddi.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     20 Feb 15 12:02 libwaspexpr.so.04 ->
libwaspexpr.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 279784 Feb 15 12:02 libwaspexpr.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     22 Feb 15 12:02 libwasphalite.so.04
-> libwasphalite.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 779872 Feb 15 12:02 libwasphalite.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     19 Feb 15 12:02 libwasphit.so.04 ->
libwasphit.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 341408 Feb 15 12:02 libwasphit.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     20 Feb 15 12:02 libwasphive.so.04 ->
libwasphive.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001  48696 Feb 15 12:02 libwasphive.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     20 Feb 15 12:02 libwaspjson.so.04 ->
libwaspjson.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 349528 Feb 15 12:02 libwaspjson.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     19 Feb 15 12:02 libwasplsp.so.04 ->
libwasplsp.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 472376 Feb 15 12:02 libwasplsp.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     21 Feb 15 12:02 libwaspsiren.so.04 ->
libwaspsiren.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 146600 Feb 15 12:02 libwaspsiren.so.4.0.3
lrwxrwxrwx 1 ir-bala2 ukaea-ap001     19 Feb 15 12:02 libwaspson.so.04 ->
libwaspson.so.4.0.3
-rwxr-xr-x 1 ir-bala2 ukaea-ap001 365384 Feb 15 12:02 libwaspson.so.4.0.3

Not sure where i am going wrong.
Kind regards,
Arun

**Reply by GiudGiud** [reply]
umm.
Ok maybe let's set WASP_DIR to `/home/ir-bala2/moose_dev/moose/framework/contrib/wasp`
or this one `/home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install` I dont remember

**Reply by abarun22** [reply]
Well, it stays the same when i set it manually.on the terminal
The correct one is /home/ir-bala2/moose_dev/moose/framework/contrib/wasp/install
and that must have been set when it's built with that script.

**Comment by abarun22** [comment]
Adding `LibtorchTorchScriptNeuralNetTest` in to the framework sources and modifying the app name to "MooseApp" helps registering the object in a right manner.

**Reply by GiudGiud** [reply]
This should always work but it s not the 'clean' solution.
You should be able to add objects in your app, registered to your app's name (exactly)
